trimesh/exchange/gltf.py

```diff
@@ -377,6 +380,34 @@ def _uri_to_bytes(uri, resolver):
     return base64.b64decode(uri[index + 7:])
 
 
+def _od_append(ordered, data):
+    """
+    Append data to an existing OrderedDict and
+    pad it to a 4-byte boundary.
+
+    Parameters
+    ----------
+    od : collections.OrderedDict
+      Keyed like { hash : data }
+    data : bytes
+      To be stored
+
+    Returns
+    ----------
+    index : int
+      Index of buffer_items stored in
+    """
+    # native python hashing is pretty fast
+    hashed = hash(data)
```

Comment:
@mikedh this looks like a very clean approach. I'm curious though, any sense of whether we're sacrificing significant performance by allowing all meshes to be converted to buffers before deduping them?

Reply:
Yeah I think it's sacrificing some performance for the sake of definitely-not-wrong bookkeeping. The bulk of the work in that function is the `tobytes` call, although it is interesting when called repeatedly (1000 times) the dominant/slowest thing are actually the `mesh.faces.max()` calls.  

That being said it is *slightly*  faster. I also discovered [this bug](https://github.com/mikedh/trimesh/pull/1182/commits/80e16e2287e5c60740603b7de72d0883733d241d) which was tripling the export time on this dummy example. 


Master, with camera bug fixed:
```
1.201 <module>  ded.py:1
└─ 1.201 export  trimesh/scene/scene.py:771
   └─ 1.201 export_scene  trimesh/exchange/export.py:188
      └─ 1.200 export_gltf  trimesh/exchange/gltf.py:62
         ├─ 0.906 _create_gltf_structure  trimesh/exchange/gltf.py:414
         │  ├─ 0.877 _append_mesh  trimesh/exchange/gltf.py:520
         │  │  ├─ 0.355 _amax  numpy/core/_methods.py:37
         │  │  │     [4 frames hidden]  numpy, <built-in>
         │  │  │        0.346 [self]  
         │  │  ├─ 0.221 _amin  numpy/core/_methods.py:41
         │  │  │     [2 frames hidden]  numpy
         │  │  ├─ 0.075 TrackedArray.tobytes  <built-in>:0
         │  │  │     [2 frames hidden]  <built-in>
         │  │  ├─ 0.072 [self]  
         │  │  ├─ 0.067 faces  trimesh/base.py:278
         │  │  │  └─ 0.055 empty  <built-in>:0
         │  │  │        [2 frames hidden]  <built-in>
         │  │  ├─ 0.049 vertices  trimesh/base.py:427
         │  │  │  └─ 0.027 empty  <built-in>:0
         │  │  │        [2 frames hidden]  <built-in>
         │  │  └─ 0.016 TrackedArray.astype  <built-in>:0
         │  │        [2 frames hidden]  <built-in>
         │  └─ 0.029 to_gltf  trimesh/scene/transforms.py:133
         ├─ 0.272 bytes.join  <built-in>:0
         │     [2 frames hidden]  <built-in>
         └─ 0.017 jsonify  trimesh/util.py:1690
            └─ 0.017 dumps  json/__init__.py:183
                  [4 frames hidden]  json


acc: 2000 buff: 2000
buffer bytes 737082240
```

This branch, with camera bug fixed:
```
0.986 <module>  ded.py:1
└─ 0.986 export  trimesh/scene/scene.py:771
   └─ 0.986 export_scene  trimesh/exchange/export.py:188
      └─ 0.985 export_gltf  trimesh/exchange/gltf.py:62
         ├─ 0.971 _create_gltf_structure  trimesh/exchange/gltf.py:445
         │  ├─ 0.938 _append_mesh  trimesh/exchange/gltf.py:551
         │  │  ├─ 0.244 _amax  numpy/core/_methods.py:37
         │  │  │     [2 frames hidden]  numpy
         │  │  │        0.242 [self]  
         │  │  ├─ 0.243 _od_append  trimesh/exchange/gltf.py:383
         │  │  │  └─ 0.235 hash  <built-in>:0
         │  │  │        [2 frames hidden]  <built-in>
         │  │  ├─ 0.185 _amin  numpy/core/_methods.py:41
         │  │  │     [2 frames hidden]  numpy
         │  │  ├─ 0.086 TrackedArray.astype  <built-in>:0
         │  │  │     [2 frames hidden]  <built-in>
         │  │  ├─ 0.065 kind  trimesh/visual/color.py:111
         │  │  │  └─ 0.065 _verify_crc  trimesh/visual/color.py:352
         │  │  │     └─ 0.065 __len__  trimesh/caching.py:498
         │  │  │        └─ 0.065 verify  trimesh/caching.py:398
         │  │  │           └─ 0.063 crc  trimesh/visual/color.py:131
         │  │  │              ├─ 0.041 fast_hash  trimesh/caching.py:651
         │  │  │              │  └─ 0.039 dict.values  <built-in>:0
         │  │  │              │        [2 frames hidden]  <built-in>
         │  │  │              └─ 0.022 crc  trimesh/base.py:263
         │  │  │                 └─ 0.022 fast_hash  trimesh/caching.py:651
         │  │  │                    └─ 0.021 <genexpr>  trimesh/caching.py:661
         │  │  │                       └─ 0.021 _xxhash  trimesh/caching.py:205
         │  │  ├─ 0.057 TrackedArray.tobytes  <built-in>:0
         │  │  │     [2 frames hidden]  <built-in>
         │  │  ├─ 0.021 [self]  
         │  │  └─ 0.020 faces  trimesh/base.py:278
         │  └─ 0.029 to_gltf  trimesh/scene/transforms.py:133
         └─ 0.014 jsonify  trimesh/util.py:1690
            └─ 0.013 dumps  json/__init__.py:183
                  [4 frames hidden]  json


acc: 2000 buff: 2
buffer bytes 368664

```

Reply:
De-duplicating the accessors with similar logic gets a pretty solid speedup:

```
0.275 <module>  ded.py:1
└─ 0.275 export  trimesh/scene/scene.py:771
   └─ 0.275 export_scene  trimesh/exchange/export.py:188
      └─ 0.275 export_gltf  trimesh/exchange/gltf.py:64
         ├─ 0.267 _create_gltf_structure  trimesh/exchange/gltf.py:497
         │  ├─ 0.237 _append_mesh  trimesh/exchange/gltf.py:606
         │  │  ├─ 0.086 _acc_append  trimesh/exchange/gltf.py:413
         │  │  │  ├─ 0.061 _xxhash  trimesh/caching.py:205
         │  │  │  ├─ 0.017 dumps  json/__init__.py:183
         │  │  │  │     [8 frames hidden]  json, <built-in>
         │  │  │  └─ 0.004 [self]  
         │  │  ├─ 0.048 TrackedArray.astype  <built-in>:0
         │  │  │     [2 frames hidden]  <built-in>
         │  │  ├─ 0.044 _od_append  trimesh/exchange/gltf.py:385
         │  │  │  ├─ 0.039 fast_hash  trimesh/caching.py:711
         │  │  │  └─ 0.004 [self]  
         │  │  ├─ 0.021 TrackedArray.tobytes  <built-in>:0
         │  │  │     [2 frames hidden]  <built-in>
         │  │  ├─ 0.018 [self]  
         │  │  ├─ 0.011 faces  trimesh/base.py:278
         │  │  │  ├─ 0.007 empty  <built-in>:0
         │  │  │  │     [2 frames hidden]  <built-in>
         │  │  │  └─ 0.003 [self]  
         │  │  └─ 0.004 __array_finalize__  trimesh/caching.py:136
         │  │     └─ 0.003 [self]  
         │  └─ 0.029 to_gltf  trimesh/scene/transforms.py:133
         │     ├─ 0.012 [self]  
         │     ├─ 0.008 _amax  numpy/core/_methods.py:37
         │     │     [3 frames hidden]  numpy, <built-in>
         │     └─ 0.003 ndarray.reshape  <built-in>:0
         │           [2 frames hidden]  <built-in>
         └─ 0.008 jsonify  trimesh/util.py:1690
            └─ 0.008 dumps  json/__init__.py:183
                  [4 frames hidden]  json


acc: 2 buff: 2
buffer bytes 368664

```

Reply:
👏 sounds good to me! The reduced size should be great. It may help with write speeds and will definitely help with transfer. Same thing for reducing the number of accessors, looks like can greatly reduce the size of the json.

I may be trying to squeeze more performance out of this then is worth the effort, but worth asking. Any chance it makes sense try a memoized version of `tobytes()`?

Reply:
Yeah agreed, I think for scenes with a bunch of instances this should be nice. I don't think I'd touch `tobytes`, as it's a pretty tiny fraction of that profile (0.021 / .275s), and it's the thing that lets us run the hash to be confident things are actually identical.